planning/MartenV4.md

```diff
@@ -0,0 +1,114 @@
+
+What could be generated in the future:
+
+* Document metadata properties -- but sad trombone, that might have to stay with Expressions if the setters are internal/private :/
+* Much more of the `ISelector` implementations, especially since there's going to be more variability when we do the document metadata
+* Finer-grained manipulation of the `IIdentityMap` 
+
+
+## Unit Testing Approach
+
+If we introduce the runtime code generation back into Marten, that's unfortunately a non-trivial "cold start" testing issue. To soften that, I suggest we get a lot more aggressive with reusable [xUnit.Net class fixtures](https://xunit.net/docs/shared-context) between tests to reuse generated code between tests, cut way down on the sheer number of database calls by not having to frequently check the schema configuration, and other `DocumentStore` overhead. 
+
+A couple other points about this:
+
+* We need to create more unique document types so we're not having to use different configurations for the same document type. This would enable more reuse inside the testing runtime
+* Be aggressive with separate schemas for different configurations
+* We could possibly turn on xUnit.net parallel test running to speed up the test cycles
+
+
+## Document Metadata
+
+* Use the configuration and tests from [pull request for Document Metadata](https://github.com/JasperFx/marten/pull/1364), but use the Lamar-backed dynamic code generation from the previous section to pull this off.
+* I think we could expand the document metadata to allow for user defined properties like "user id" or "transaction id" much the same way we'll do for the EventStore metadata. We'd need to think about how we extend the document tables and how metadata is attached to a document session
+
+
+## Project / Assembly Restructuring
+
+* Pull everything to do with Schema object generation and difference detection to a separate library (`IFeatureSchema`, `ISchemaObject`, etc.). Mostly to clean out the main library, but also because this code could easily be reused outside of Marten. Separating it out might make it easier to test and extend that functionality, which is something that occasionally gets requested. There's also the possibility of further breaking that into abstractions and implementations for the long run of getting us ready for Sql Server or other database engine support
+
+* *Possibly* pull the ADO.Net helper code like `CommandBuilder` and the extension methods into a small helper library somewhere else (I'm nominating the [Baseline](https://jasperfx.github.io/baseline) repository). This code is copied around to other projects as it is, and it's another way of getting stuff out of the main library and the test suite.
+
+* More in the event store section, but let's get the event store functionality into its own library (or possibly a couple other libraries for async projection support).
+
+## F# Improvements
+
+WE NEED TO HAVE AN F# SUBCOMMITTEE TO HELP HERE. 
```

Comment:
I was talking with @TheAngryByrd some time ago. He was open for taking the Marten.FSharp. I think that we should start with that. For me making the Marten F# friendly is one of the goals. F# community seems to be aligned with the Marten vision.
We should start from providing support for Discriminated Unions. https://github.com/JasperFx/marten/issues/1283
I was also thinking about using a sample from Domain Modeling Made functional and making sure that we can support the presented approach https://github.com/swlaschin/DomainModelingMadeFunctional

Reply:
Chet and @wastaz too have all said they'd be happy to help. I think the `IField` revamp makes things like the discriminated unions more feasible

Reply:
cc @baronfel

Reply:
Maybe we could schedule some call to discuss the initial plan and the biggest pain points?

Reply:
Yes let's set something up for later this week.  Thursday-Sunday EDT I'm good with.

Reply:
I'm down for that. My workload is suddenly much lower than normal, so I'm game. I can do much earlier in the day with the shelter in place orders.

I'm UTC-5 timewise.

Reply:
I'm UTC+2. Normally I have meetings between 12-3PM UTC time, but I should be flexible.

Reply:
Im also in UTC+2. Im normally pretty busy during work hours, and I have a tiny ~toddler~ terror that gets put to bed somewhere between 18-19 so Im sadly a little bit less flexible. It would be nice to join, but if we cant figure out a time that works for me Im not gonna be offended if I get left behind. In that case Ill try to communicate my thoughts in writing.

Reply:
@wastaz since you're the most restricted, can you propose some times?

Reply:
Sure, I can try. :)
I think for me the easiest times that I can fit in a meeting (while not having to fight off the child who wants to crash every meeting I have if he's home when I am) would either be trying to fit it in while he's still in preschool or in the evening after he's in bed.

Im feeling bad for suggesting this time given that jeremy is in -5 (but otoh he seems awfully much like a morning person given how often he answers on gitter when Im expecting him to be asleep), but around 12:00 UTC would probably work thu/fri and give me about 1,5h before I have to run off. Otherwise I am pretty sure that I can fit in 18:00 UTC on any of thu-sun.

Reply:
12 UTC on Friday works for me. 18 UTC Thu-Fri also.

Reply:
I'm definitely the most flexible and fine with either of those times. Let's try for 12 UTC Friday?

Reply:
You know what guys. Im an idiot. Im really sorry. I completely missed that its international workers day on friday so my kid is not going to preschool, so he's gonna be home and awake and 12UTC is prime time for climbing on dad and jumping around and being everywhere at once.

Reply:
So maybe 18:00 UTC on Thursday? @jeremydmiller @TheAngryByrd  @wastaz?

Reply:
Unfortunately that's the one time that doesn't work for me here.  I now have work related meetings from 18:00 UTC to 20:00 UTC on Thursday.

Reply:
I'm in the same boat as @TheAngryByrd in terms of meetings/timing.

Reply:
So Maybe Thursday at 20 UTC or Friday at 20 UTC?

Reply:
Those work for me.

Reply:
Its a bit on the late end for me honestly (my child is a "morning" person, I...am not, but actually going to bed in time in the evenings helps me keep my sanity). But I guess I might be able to be convinced to go for that time, but Ill probably drop off by 21UTCish. Also friday would probably work better than thursday.

Reply:
Ok meet in the gitter and figure out what techi to use, skype/zoom/gotomeeting/etc?

Reply:
I can set up the hangout link. @jeremydmiller prefers the zoom usually. @jeremydmiller @wastaz @TheAngryByrd are we meeting today or tomorrow? I'd like to know as 20UTC it's also kinda late for me, so I need to know to not fell asleep on the couch 😉

Reply:
I can do either.

Reply:
@jeremydmiller could you suggest some time suitablle for you? I think that meeting without you won't make much sense ;)